libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,269 @@
+    """Use for images / audio / video"""
+
+    data: bytes
+    encoding: str  # base64, utf-8, URI, etc.
+    mimetype: str  # image/png, text/plain, etc.
+
+
+# This is likely not needed during indexing right now
+# class MultiContent(Content):
+#     """Group multiple pieces of content together"""
+#     contents: List[Content]
+#     metadata: dict
+
+
+T = TypeVar("T", bound=IndexableData)
+
+
+class UpsertResponse(TypedDict):
+    """An indexing result."""
+
+    failed: Sequence[str]
+    indexed: Sequence[str]
+
+
+class DeleteResponse(TypedDict):
+    """A response to a delete request."""
+
+    num_deleted: NotRequired[int]
+    num_failed: NotRequired[int]
+    failed: NotRequired[Sequence[str]]
+    deleted: NotRequired[Sequence[str]]
+
+
+class Sort(TypedDict):
+    """A sort object."""
+
+    field: str
+    order: NotRequired[Literal["asc", "desc"]]  # Assume asc by default
+
+
+class FilterQuery(TypedDict):
+    """Query for an item.
+
+    This enables querying for an item using similarity search +
+    standard operations on the relational data associated with the item.
+    """
+
+    filter: Optional[dict]  # We'll need to type this # <-- Could stand for raw filter?
+    limit: Optional[int]  # equivalent to top-k right now
+    offset: Optional[int]  # e.g., interpret as page offset (maybe a better name?)
+    # e.g., browse through documents by publication date
+    sort: Optional[List[Sort]]  # Optional multiple sort fields
+
+
+C = TypeVar("C")
+
+
+Vector = List[float]
+
+
+class BaseIndex(Generic[T]):
+    """An index represent a collection of items that can be queried.
+
+    The index is responsible for:
+
+    1. Storing the content
+    2. Allowing retrieval of the content by id
+    3. Supporting queries against the metadata associated with the content
+
+    The index is **NOT** responsible for:
+
+    1. Support search queries against the content.
+
+    The types of supported queries seems very dependent on the actual
+    """
+
+    @abc.abstractmethod
+    def upsert(self, data: Iterable[T], /, **kwargs: Any) -> Iterable[UpsertResponse]:
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Iterable[str],  # Sequence or Iterable
+        /,
+    ) -> Iterable[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],  # Sequence or Iterable
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete an item by id."""
+
+    # Delete and get are part of the READ/WRITE interface.
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def lazy_get(self, query: FilterQuery, /, **kwargs: Any) -> Iterable[T]:
```

Comment:
why prefix with lazy_ if get_by_ids isn't prefixed with lazy_

Reply:
also for usability maybe we just unpack query into separate args

Reply:
I packaged it to make it easier to roll out changes, but yeah we can unpack, and rename to get_by_query?